### DISCUSSION TITLE
Automating PDF to JPG and text recognition

### DISCUSSION THREAD

**Discussion by maimaik05** [pymupdf/PyMuPDF#2473] [Looking for help]
Hello!

I'm trying to automate a process that takes a PDF, converts it to a JPG, reads and finds the text on the PDF, and uses the jpg with the location of the text in tableau to plot a coordinate. When doing this manually through Adobe, I would take the pdf, export to a postscript file, open that as a pdf, then export to a jpg. I want to try and automate this process so I'm using pymupdf to convert the pdf to a jpg currently.

I have the script working but the coordinates are not in line with the JPG, and I I'm not sure what the problem is. I use PDFMiner to do the text recognition on the pdf, does that have something to do with it? I realize pyMuPDF has text recognition abilities, but I am building off of someone else's work so I would prefer not to change too much of it if possible! 

I'm not sure if it could be that the postscript flattened the pdf and without that step its not sizing up correctly? I did try doc.save('filename', garbage = 4,deflate = True) but that didnt change anything. It's been difficult to pinpoint the issue here as you can see.

I can send the code privately if needed, if that's ok.
Also this is my first time posting anything here, let me know if you need more info/detail.

**Comment by JorjMcKie** [comment]
> but the coordinates are not in line with the JPG

Maybe we end up looking at a file example, but let me first ask what the problem is specifically here:
Not matching at all (completely some place else) or just a bit mispositioned?

I can't recall right now which coordinate system pdfminer is actually using (PDF?).

When you extract text together with its coordinates on the PDF page, then you will be dealing with points (1 inch = 72 points) and these are floats. The PDF standard coordinate system uses the page's bottom-left point as (0, 0). And maybe pdfminer is doing this too.
**_Image_** addressing in contrast works with pixels using integers, and point (0, 0) is the top-left corner.

PyMuPDF has several "geometry" objects like **_rectangles_** and **_matrices_** to deal with all that. Among other things, there is a function that computes a coordinate system transformation matrix to convert e.g. PDF page coordinates to corresponding coordinates inside an image that was rendered from that page.
Potentially that might help you.

**Reply by maimaik05** [reply]
It looks like the coordinates are shifted over, so it looks correct in relation to each other, but not in the correct location. I can send over a picture tomorrow. 

The script finds the text box and we calculate the center of the box to get the coordinate in pixels. I think it uses PDFParser?

I think that function might help. When we converted the pdfs manually with adobe we used the pdfminer portion of the script to plot the coordinates from there, and we had no problems with that. Do you have an example of that function?

Thank you!

**Reply by JorjMcKie** [reply]
Well, a complete solution in PyMuPDF would go along this line:
1. Determine the bbox of your text. This could be the outcome of a text search, or be part of a (complete) text extraction from the page, or similar.
2. Make an image of the page. This rendering generates an internal image format called "Pixmap", which can be saved in a number of output formats like PNG, JPEG and more. When making the pixmap this is the time you decide on the output quality by providing a desired DPI value. That resolution also directly determines the image's width and height and hence the coordinate values of everything on it.
3. Make a transformation matrix which converts page coordinates to image coordinates.

A more or less complete snippet:
```python
import fitz # PyMuPDF
doc = fitz.open("input.pdf")
page = doc[0]   # some page in the PDF (0-based page numbers)
pix = page.get_pixmap(dpi=150)  # render page to image
matrix = page.rect.torect(pix.irect)  # matrix converts page to image coordinates
needle = "search this text"  # assume we are looking for some text

rect_list = page.search_for(needle)  # find all text occurrences on page

# print each text location on the image
print(f"Text '{needle}' occurs here on the image:")
for rect in rect_list:
    img_rect = (rect * matrix).irect  # image coordinates have integer coordinates
    print(rect, "==>", img_rect)  # outputs all rectangles of needle on image
```

**Reply by maimaik05** [reply]
Thank you!

Could we save the conversion with the matrix to a pdf file after that?

Something like this maybe?
Doc=fitz.open(inputpdf)
Page=doc[0]
...
Page.rect.torect(pix.irect)
Doc.save(outputpdf)

**Reply by JorjMcKie** [reply]
Won't work like that.
I suggest you look at the source code of method `.torect()` in the `Rect` class and try to replicate that logic in whatever your preferred environment is.
It is Python code, it should be possible to do that with reasonable effort.

**Reply by maimaik05** [reply]
I'm trying to find the source code, but can't seem to locate it. Which folder is it in?

**Reply by JorjMcKie** [reply]
Look [here](https://github.com/pymupdf/PyMuPDF/blob/main/fitz/helper-geo-py.i), line 512.

**Reply by maimaik05** [reply]
I'm sorry, I'm having trouble seeing how I can use the returned matrix to save it as a pdf. Can it be used in one of the parameters when using Document.save()?

**Reply by JorjMcKie** [reply]
No. But you can compute it as described and store it's 6 values as a JSON file.
As I wrote, that matrix transform document page coordinates to image coordinates. If all pages of the document have the same width/height, the so have the produced images and you can use the same matrix for all pages of the document.
In you application, read the JSON file (which will be 6 floats) and use it for coordinate computation.

**Comment by maimaik05** [comment]
I figured out what the issue was! In PDFMiner, the X and Y coordinates for the page size are reverse of what PyMuPDF uses (the page.rect with PyMuPDF is (0,0,3168,2448) but with PDFMiner the page.mediabox is (0,0,2448,3168). Thank you for all the help, I really appreciate it
